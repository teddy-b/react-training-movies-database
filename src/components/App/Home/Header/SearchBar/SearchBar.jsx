/* @flow */

import * as React from 'react'
import { Link } from 'react-router-dom'

import {
  StyledSearchBar,
  StyledSearchBarTitle,
  StyledSearchInput,
  StyledSearchFooter,
  StyledSearchBy,
  StyledSearchByBtns,
  StyledSearchByBtn,
  StyledSearchBtn
} from './StyledSearchBar'
import Logo from '../../../../shared/Logo'
import StyledLink from '../../../../shared/styled/StyledLink'
import { SEARCH_BY, GENRES } from '../../../../../constants/global'

type Props = {
  onSearch: (searchText: string, searchBy: string) => void,
  onSearchMoviesByGenre: () => void,
  onSearchMoviesByTitle: () => void,
  searchBy: string,
}

type State = {
  searchText: string,
}

class SearchBar extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      searchText: ''
    }
  }

  render() {
    const {
      onSearch,
      onSearchMoviesByGenre,
      onSearchMoviesByTitle,
      searchBy
    } = this.props

    return (
      <StyledSearchBar>
        <StyledLink to="/">
          <Logo />
        </StyledLink>
        <StyledSearchBarTitle>Find your movie</StyledSearchBarTitle>
        <StyledSearchInput
          placeholder="Start typing movie title or genre"
          type="text"
          onChange={event => this.setState({
            searchText: event.target.value
          })}
          list="genres"
        />
        {searchBy === SEARCH_BY.genre &&
          <datalist id="genres">
            {GENRES.map(genre => <option key={genre} value={genre} />)}
          </datalist>
        }
        <StyledSearchFooter>
          <StyledSearchBy>Search by </StyledSearchBy>
          <StyledSearchByBtns>
            <StyledSearchByBtn
              selected={searchBy === SEARCH_BY.title}
              onClick={onSearchMoviesByTitle}
            >
              Title
            </StyledSearchByBtn>
            <StyledSearchByBtn
              selected={searchBy === SEARCH_BY.genre}
              onClick={onSearchMoviesByGenre}
            >
              Genre
            </StyledSearchByBtn>
          </StyledSearchByBtns>
          <Link to={`/search/${searchBy}/${this.state.searchText}`}>
            <StyledSearchBtn
              onClick={() => onSearch(this.state.searchText, searchBy)}
            >
              Search
            </StyledSearchBtn>
          </Link>
        </StyledSearchFooter>
      </StyledSearchBar>
    )
  }
}

export default SearchBar
