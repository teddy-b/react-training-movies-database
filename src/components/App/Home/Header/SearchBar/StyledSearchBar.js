import styled from 'styled-components'

import { black, grey2, grey4, red, white } from '../../../../../constants/styles'

export const StyledSearchBar = styled.div`
  background-color: ${grey4};
  color: ${white};
  padding: .5em;
`

export const StyledSearchBarTitle = styled.p`
  text-transform: uppercase;
`

export const StyledSearchInput = styled.input`
  width: 98%;
  padding: .8em;
  background-color: ${black};
  color: inherit;
  border: none;
  border-bottom: 1px solid ${red};
`

export const StyledSearchFooter = styled.div`
  padding: .5em;
  display: flex;
  justify-content: space-between;
  align-items: center;
`

export const StyledSearchBy = styled.div`
  text-transform: uppercase;
`

export const StyledSearchByBtns = styled.div`
  flex-grow: 2;
`

export const StyledSearchByBtn = styled.button`
  color: ${white};
  border: none;
  cursor: pointer;
  margin: .3em;
  padding: .6em;
  background-color: ${props => (props.selected ? red : grey2)};
`

export const StyledSearchBtn = styled.button`
  color: ${white};
  border: none;
  cursor: pointer;
  margin: .2em;
  padding: .8em;
  width: 15vw;
  background-color: ${red};
`
