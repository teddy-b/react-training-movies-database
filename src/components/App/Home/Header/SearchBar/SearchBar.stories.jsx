/* @flow */

import * as React from 'react'
import { MemoryRouter } from 'react-router-dom'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import type { Renderable } from '@storybook/react'

import SearchBar from './SearchBar'
import { SEARCH_BY } from '../../../../../constants/global'

const props = {
  onSearch: action('clicked'),
  onSearchMoviesByGenre: action('clicked'),
  onSearchMoviesByTitle: action('clicked'),
  searchBy: SEARCH_BY.title
}

storiesOf('Components/SearchBar', module)
  .add('search by title', (): Renderable => (
    <MemoryRouter>
      <SearchBar
        {...props}
        searchBy={SEARCH_BY.title}
      />
    </MemoryRouter>
  ))
  .add('search by genre', (): Renderable => (
    <MemoryRouter>
      <SearchBar
        {...props}
        searchBy={SEARCH_BY.genre}
      />
    </MemoryRouter>
  ))
