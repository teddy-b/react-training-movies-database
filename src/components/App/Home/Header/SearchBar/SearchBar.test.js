import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'

import SearchBar from './SearchBar'

jest.mock('../../../../shared/Logo', () => 'Logo')

describe('SearchBar', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <SearchBar
        onChangeSearchParam={jest.fn()}
        onSearch={jest.fn()}
      />
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
  
  it('should trigger onChangeSearchParam onClick', () => {
    const onChangeSearchParamMock = jest.fn()
    const component = shallow(
      <SearchBar
        onChangeSearchParam={onChangeSearchParamMock}
        onSearch={jest.fn()}
      />
    )
  
    component.find('button').at(0).simulate('click')
  
    expect(onChangeSearchParamMock.mock.calls.length).toBe(1)
  
    component.find('button').at(1).simulate('click')
  
    expect(onChangeSearchParamMock.mock.calls.length).toBe(2)
  })
  
  it('should trigger onChangeSearchParam onClick', () => {
    const onSearchMock = jest.fn()
    const component = mount(
      <SearchBar
        onChangeSearchParam={jest.fn()}
        onSearch={onSearchMock}
      />
    )
  
    component.find('input').at(0).simulate('change', { target: { value: 'New value' } })
  
    component.find('button').at(2).simulate('click')
  
    expect(onSearchMock).toBeCalled()
  })
})
