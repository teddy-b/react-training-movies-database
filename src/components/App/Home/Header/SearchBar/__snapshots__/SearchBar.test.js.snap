// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders correctly 1`] = `
<div
  className="searchBar"
>
  <Logo />
  <p
    className="searchBarTitle"
  >
    Find your movie
  </p>
  <input
    className="searchInput"
    type="text"
  />
  <div
    className="searchFooter"
  >
    <div
      className="searchBy"
    >
      Search by 
    </div>
    <div
      className="searchByBtns"
    >
      <button
        className="searchByTitleBtn"
        onClick={[Function]}
      >
        Title
      </button>
      <button
        className="searchByGenreBtn"
        onClick={[Function]}
      >
        Genre
      </button>
    </div>
    <button
      className="searchBtn"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;
